src/planscape/forsys/views.py

```diff
@@ -3,43 +3,321 @@
 def convert_rdf_to_pddf(rdf):
   pddf = pd.DataFrame.from_dict({ key : np.asarray(rdf.rx2(key)) for key in rdf.names })
   return pddf
 
+# Converts dictionary of lists to R dataframe.
+# The lists must have equal length.
+def convert_dictionary_of_lists_to_rdf(lists):
+  data = {}
+  for key in lists.keys():
+    if len(lists[key]) == 0:
+      continue
+    el = lists[key][0]
+    if isinstance(el, str):
+      data[key] = rpy2.robjects.StrVector(lists[key])
+    elif isinstance(el, float):
+      data[key] = rpy2.robjects.FloatVector(lists[key])
+    elif isinstance(el, int):
+      data[key] = rpy2.robjects.IntVector(lists[key])
+
+  rdf = rpy2.robjects.vectors.DataFrame(data)
+  return rdf
+
+# Translates polygon.
+def translate_polygon(polygon, dx, dy):
+  coords = []
+  for ring in polygon.coords:
+    for point in ring:
+      coords.append((point[0] + dx, point[1] + dy))
+  translated = Polygon( tuple(coords) ) 
+  translated.srid = polygon.srid
+  return translated
+
+def get_boundary_debug_info(boundaries, project_area):
+  boundary_response = []
+  for b in boundaries:
+    boundary_response.append("%s (id=%s, intersection area=%f)"%(b.shape_name, b.boundary_id, b.geometry.intersection(project_area).area))
+  return boundary_response
+
+def get_raster_debug_info(rasters):
+  raster_response = []
+  for c in rasters:
+    r = rasters[c]
+    d = r.bands[0].data()
+    count = np.count_nonzero(~np.isnan(d))
+    mean = np.sum(d[~np.isnan(d)]) / count
+    shape = np.shape(d)
+    raster_response.append("%s (non-nan area: %d, mean: %f, shape: %d x %d)"%(c, count, mean, shape[0], shape[1]))
+  return raster_response
+
+def get_condition_rasters(condition, region):
+  conditions = BaseCondition.objects.filter(condition_name=condition).filter(region_name=region).all()
+  if len(conditions) == 0:
+    raise LookupError("no condition with name, %s, exists in region, %s"%(condition, region))
+  if len(conditions) > 1:
+    raise LookupError("more than 1 condition with name, %s, exists in region, %s"%(condition, region))
+  c = conditions[0]
+  condition_files = Condition.objects.filter(condition_dataset_id=c.id).all()
+  if len(condition_files) == 0:
+    raise LookupError("no condition filename exists for condition id, %d"%(c.id))
+  if len(condition_files) > 1:
+    raise LookupError("more than 1 condition filename exists for condition id, %d"%(c.id))
+
+  condition_rasters = ConditionRaster.objects.filter(name=condition_files[0].raster_name)
+  return condition_rasters
+
+def raster_extent_overlaps_project_area(raster, project_area):
+  e = raster.extent
+  e_polygon = Polygon( ((e[0], e[1]),
+                        (e[2], e[1]),
+                        (e[2], e[3]),
+                        (e[0], e[3]),
+                        (e[0], e[1])) )
+  return e_polygon.overlaps(project_area)
+
+# Merges add-on raster with base_raster.
```

Comment:
Please expand this comment, I still don't know what this function is for. What are add-on raster and base_raster?

Reply:
I'll add more comments.

This is for merging two rasters into a single raster.
Sometimes a project area may cover multiple DB raster chunks.
A single raster is easier to pass around and run computations on.

Reply:
Done - I've added more comments.
I hope it's clearer. gdal libraries tend to refer to this operation with terms like "mosaic" and "merge".